### PR TITLE
Convert CUDA tests to use Kokkos

### PR DESCRIPTION
**Pull request by masterleinad**
This should cover all tests that are not requiring CUDA/GPU/Device-aware MPI, CUDA objects (sparse matrix, vectors, preconditioners), or CUDA MatrixFree.

### CODE REVIEW

File: include/deal.II/lac/affine_constraints.templates.h [hunk 1]
```diff
@@ -2285,22 +2273,30 @@ namespace internal
...
+      Kokkos::deep_copy(constrained_local_dofs_device,
+                        Kokkos::View<size_type *, Kokkos::HostSpace>(
+                          constrained_local_dofs_host.data(),
+                          constrained_local_dofs_host.size()));
+
+      using ExecutionSpace =
+        MemorySpace::Default::kokkos_space::execution_space;
+      ExecutionSpace exec;
+      auto           local_values = vec.get_values();
+      Kokkos::parallel_for(
```
Comment (Rombur) [inline include/deal.II/lac/affine_constraints.templates.h:2291]: name

File: tests/base/kokkos_point.cc [hunk 2]
```diff
@@ -19,84 +19,74 @@
...
 {
-  Number *           check;
   const unsigned int n_tests = 16;
 
-  auto cuda_error = cudaMalloc(&check, n_tests * sizeof(Number));
-  AssertCuda(cuda_error);
+  Kokkos::View<Number *, MemorySpace::Default::kokkos_space> check("check",
+                                                                   n_tests);
 
   // Miscellaneous
```
Comment (Rombur) [inline tests/base/kokkos_point.cc:31]: What does this comment mean?

File: tests/base/kokkos_point.cc [hunk 3]
```diff
@@ -19,84 +19,74 @@
...
+      check[12] = s_1;
+      auto s_2  = p_2.square();
+      check[13] = s_2;
+      auto s_3  = p_3.distance(p_5);
+      check[14] = s_3;
+      auto s_4  = p_4.distance_square(p_1);
+      check[15] = s_4;
+    });
+
+  exec.fence();
```
Comment (Rombur) [inline tests/base/kokkos_point.cc:86]: Unnecessary fence

File: tests/base/kokkos_tensor_01.cc [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    Kokkos::MDRangePolicy<ExecutionSpace, Kokkos::Rank<2>>(exec,
+                                                           {{0, 0}},
+                                                           {{dim, dim}}),
+    KOKKOS_LAMBDA(int i, int j) { t_dev()[i][j] = j + i * dim + 1.; });
+  Kokkos::parallel_for(
+    Kokkos::RangePolicy<ExecutionSpace>(exec, 0, 1), KOKKOS_LAMBDA(int) {
+      norm_dev()        = t_dev().norm();
+      norm_square_dev() = t_dev().norm_square();
+    });
+  exec.fence();
```
Comment (Rombur) [inline tests/base/kokkos_tensor_01.cc:68]: unnecessary fence

File: tests/base/kokkos_tensor_02.cc [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+template <int rank, int dim, typename Number>
+struct InitFunctor;
+
+template <int dim, typename Number>
+struct InitFunctor<0, dim, Number>
+{
+  Kokkos::View<Tensor<0, dim, Number>, MemorySpace::Default::kokkos_space> t;
+
+  KOKKOS_FUNCTION void
+  operator()(int k) const
```
Comment (Rombur) [inline tests/base/kokkos_tensor_02.cc:31]:
```suggestion
  operator()(int ) const
```

File: tests/lac/affine_constraints_set_zero.cc [hunk 6]
```diff
@@ -56,15 +47,23 @@ test()
...
-      ghosted.get_values(), numproc, myid * numproc);
+    auto ghosted_values = ghosted.get_values();
+
+    Kokkos::parallel_for(
+      Kokkos::RangePolicy<ExecutionSpace>(exec, 0, numproc),
+      KOKKOS_LAMBDA(int i) {
+        int offset        = myid * numproc;
+        ghosted_values[i] = 1.0 + i + offset;
+      });
+    exec.fence();
```
Comment (Rombur) [inline tests/lac/affine_constraints_set_zero.cc:66]: Do we need this fence?

File: tests/lac/affine_constraints_set_zero.cc [hunk 7]
```diff
@@ -76,15 +75,22 @@ test()
...
-      distributed.get_values(), numproc, myid * numproc);
+    auto distributed_values = distributed.get_values();
+
+    Kokkos::parallel_for(
+      Kokkos::RangePolicy<ExecutionSpace>(exec, 0, numproc),
+      KOKKOS_LAMBDA(int i) {
+        int offset            = myid * numproc;
+        distributed_values[i] = 1.0 + i + offset;
+      });
+    exec.fence();
```
Comment (Rombur) [inline tests/lac/affine_constraints_set_zero.cc:92]: Same as above? I would think that `compress` should be in charge of fencing.

**Comment by masterleinad** [comment]
Since we are not using any shared memory space like `CudaUVM`, only the default execution space instance explicitly and only the two-argument `deep_copy` (that fences for and after all execution space instances), we should be fine without ever fencing explicitly. That being said, I think we should reevaluate the fences after profiling.

**Comment by Rombur** [comment]
>  we should be fine without ever fencing explicitly. That being said, I think we should reevaluate the fences after profiling.

Why? If there are not necessary, why do you want to add them?

**Comment by masterleinad** [comment]
> Why? If they are not necessary, why do you want to add them?

I think that it makes sense to at least avoid global fences in certain situations (which we will need to do some more profiling for) and I would rather have the implementation ready for that. If the global fences don't hurt us, adding some more fences is likely also not performance-critical. Either way, I think we should make it easy for users not to forget about synchronization and offer options to be more specific if necessary. Also, I already removed the fences you were commenting on.
I would prefer to get the conversion done and consider synchronization again later.

**Review by Rombur — APPROVED** [review]

**Pull request merged**